File: AppBox/Common/UpdateHandler/UpdateHandler.m
//
//  UpdateHandler.m
//  AppBox
//
//  Created by Vineet Choudhary on 30/12/16.
//  Copyright © 2016 Developer Insider. All rights reserved.
//

#import "UpdateHandler.h"

@implementation UpdateHandler

//MARK: - Homebrew Detection

+ (BOOL)isInstalledViaHomebrew {
    NSFileManager *fileManager = [NSFileManager defaultManager];
    // Check Apple Silicon Homebrew path
    if ([fileManager fileExistsAtPath:@"/opt/homebrew/Caskroom/appbox"]) {
        return YES;
    }
    // Check Intel Homebrew path
    if ([fileManager fileExistsAtPath:@"/usr/local/Caskroom/appbox"]) {
        return YES;
    }
    return NO;
}

//MARK: - Check for update

+ (void)showHomebrewUpdateAlert {
    NSString *command = @"brew upgrade --cask appbox";
    
    NSAlert *alert = [[NSAlert alloc] init];
    [alert setMessageText:@"New Version Available"];
    [alert setInformativeText:@"A newer version of \"AppBox\" is available.\n\nSince you installed AppBox via Homebrew, please run this command in Terminal:"];
    [alert setAlertStyle:NSAlertStyleInformational];
    [alert setAccessoryView:[self commandTextFieldWithString:command width:250]];
    [alert addButtonWithTitle:@"Copy & Close"];
    [alert addButtonWithTitle:@"Release Notes"];
    [alert addButtonWithTitle:@"Close"];
    NSModalResponse response = [alert runModal];
    if (response == NSAlertFirstButtonReturn) {
        [[NSPasteboard generalPasteboard] clearContents];
        [[NSPasteboard generalPasteboard] setString:command forType:NSPasteboardTypeString];
    } else if (response == NSAlertSecondButtonReturn) {
        [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:abGitHubLatestReleaseURL]];
    }
}

+ (void)showUpdateAlertWithUpdateURL:(NSURL *)url{
    NSString *command = @"curl -s https://getappbox.com/install.sh | bash";
    
    NSAlert *alert = [[NSAlert alloc] init];
    [alert setMessageText:@"New Version Available"];
    [alert setInformativeText:@"A newer version of \"AppBox\" is available.\n\nTo update, run this command in Terminal:"];
    [alert setAlertStyle:NSAlertStyleInformational];
	[alert setAccessoryView:[self commandTextFieldWithString:command width:400]];
    [alert addButtonWithTitle:@"Copy & Close"];
    [alert addButtonWithTitle:@"Release Notes"];
    [alert addButtonWithTitle:@"Close"];
    NSModalResponse response = [alert runModal];
    if (response == NSAlertFirstButtonReturn){
        [[NSPasteboard generalPasteboard] clearContents];
        [[NSPasteboard generalPasteboard] setString:command forType:NSPasteboardTypeString];
    } else if (response == NSAlertSecondButtonReturn) {
        [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:abGitHubLatestReleaseURL]];
    }
}

+ (NSView *)commandTextFieldWithString:(NSString *)command width:(CGFloat)width {
    NSTextField *textField = [[NSTextField alloc] initWithFrame:NSMakeRect(0, 0, width, 20)];
    [textField setStringValue:[NSString stringWithFormat:@"%@", command]];
    [textField setEditable:NO];
    [textField setSelectable:YES];
    [textField setBordered:YES];
    [textField setBezelStyle:NSTextFieldRoundedBezel];
    [textField setFont:[NSFont monospacedSystemFontOfSize:12.0 weight:NSFontWeightMedium]];
    [textField setBackgroundColor:[NSColor controlBackgroundColor]];
    [textField setAlignment:NSTextAlignmentCenter];
    return textField;
}

+ (void)showAlreadyUptoDateAlert{
    NSString *versionString = [[[NSBundle mainBundle] infoDictionary] valueForKey:@"CFBundleShortVersionString"];
    [Common showAlertWithTitle:@"You’re up-to-date!" andMessage:[NSString stringWithFormat:@"AppBox %@ is currently the newest version available.", versionString]];
}

+ (void)isNewVersionAvailableCompletion:(void (^)(bool available, NSURL *url))completion{
    @try {
        DDLogDebug(@"Checking for new version...");
        
        if ([self isInstalledViaHomebrew]) {
            // Check Homebrew API for latest available version
            [NetworkHandler requestWithURL:abHomebrewCaskAPI withParameters:nil andRequestType:RequestGET andCompletetion:^(id responseObj, NSInteger httpStatus, NSError *error) {
                @try {
                    if (error == nil && [responseObj isKindOfClass:[NSDictionary class]] &&
                        [((NSDictionary *)responseObj).allKeys containsObject:@"version"]) {
                        
                        NSString *latestVersion = [responseObj valueForKey:@"version"];
                        NSString *versionString = [[[NSBundle mainBundle] infoDictionary] valueForKey:@"CFBundleShortVersionString"];
                        if (!versionString || !latestVersion) {
                            completion(false, nil);
                            return;
                        }
                        
                        DDLogDebug(@"Current Version - %@ <=> Homebrew Latest Version - %@", versionString, latestVersion);
                        
                        NSString *cleanLatest = [self extractVersionString:latestVersion];
						NSString *cleanCurrent = [self extractVersionString:versionString];
                        NSComparisonResult result = [cleanLatest compare:cleanCurrent options:NSNumericSearch];
                        
                        completion((result == NSOrderedDescending), nil);
                    } else {
                        // Homebrew API failed — fallback to GitHub release check
                        DDLogInfo(@"Homebrew API failed, falling back to GitHub release check.");
                        [self checkGitHubReleaseWithCompletion:completion];
                    }
                }
                @catch (NSException *exception) {
                    DDLogError(@"Exception %@",exception.abDescription);
                    [self checkGitHubReleaseWithCompletion:completion];
                }
            }];
        } else {
            [self checkGitHubReleaseWithCompletion:completion];
        }
    } @catch (NSException *exception) {
        completion(false, nil);
		DDLogInfo(@"Exception %@",exception.abDescription);
    }
}

+ (void)checkGitHubReleaseWithCompletion:(void (^)(bool available, NSURL *url))completion {
    [NetworkHandler requestWithURL:abGitHubLatestRelease withParameters:nil andRequestType:RequestGET andCompletetion:^(id responseObj, NSInteger httpStatus, NSError *error) {
        @try {
            if (error == nil && [responseObj isKindOfClass:[NSDictionary class]] &&
                [((NSDictionary *)responseObj).allKeys containsObject:@"tag_name"] &&
                [((NSDictionary *)responseObj).allKeys containsObject:@"html_url"]){
                
                NSString *tag = [responseObj valueForKey:@"tag_name"];
                NSString *versionString = [[[NSBundle mainBundle] infoDictionary] valueForKey:@"CFBundleShortVersionString"];
                if (!versionString || !tag) {
                    completion(false, nil);
                    return;
                }
                
                DDLogDebug(@"Current Version - %@ <=> GitHub Latest Version - %@", versionString, tag);
                
                NSString *latestVersion = [self extractVersionString:tag];
                NSString *currentVersion = [self extractVersionString:versionString];
                NSComparisonResult result = [latestVersion compare:currentVersion options:NSNumericSearch];
                
                completion((result == NSOrderedDescending), [NSURL URLWithString:[responseObj valueForKey:@"html_url"]]);
            } else {
                completion(false, nil);
            }
        }
        @catch (NSException *exception) {
            DDLogError(@"Exception %@",exception.abDescription);
            completion(false, nil);
        }
    }];
}

+ (NSString *)extractVersionString:(NSString *)input {
    if (!input) return @"0";
    // Remove everything except digits and dots
    NSMutableString *version = [NSMutableString string];
    BOOL foundDigit = NO;
    for (NSUInteger i = 0; i < input.length; i++) {
        unichar c = [input characterAtIndex:i];
        if (c >= '0' && c <= '9') {
            [version appendFormat:@"%C", c];
            foundDigit = YES;
        } else if (c == '.' && foundDigit) {
            [version appendFormat:@"%C", c];
        }
    }
    // Remove trailing dot if any
    if (version.length > 0 && [version characterAtIndex:version.length - 1] == '.') {
        [version deleteCharactersInRange:NSMakeRange(version.length - 1, 1)];
    }
    return version.length > 0 ? [version copy] : @"0";
}


@end

File: AppBox/Common/UpdateHandler/UpdateHandler.swift
//
//  UpdateHandler.swift
//  AppBox

import Foundation
import AppKit
import AppBoxCore
import os

public final class UpdateHandler: NSObject {

    private static let log = Logger(subsystem: "com.developerinsider.AppBox", category: "UpdateHandler")
    private static let gitHubLatestReleaseURL = "https://github.com/getappbox/AppBox-iOSAppsWirelessInstallation/releases/latest"

    private static var currentVersion: String? {
        Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
    }

    public class func isInstalledViaHomebrew() -> Bool {
        return ABHomebrewDetector.isInstalledViaHomebrew()
    }

    public class func showHomebrewUpdateAlert() {
        let command = "brew upgrade --cask appbox"
        let alert = NSAlert()
        alert.messageText = "New Version Available"
        alert.informativeText = "A newer version of \"AppBox\" is available.\n\nSince you installed AppBox via Homebrew, please run this command in Terminal:"
        alert.alertStyle = .informational
        alert.accessoryView = commandTextField(command, width: 250)
        alert.addButton(withTitle: "Copy & Close")
        alert.addButton(withTitle: "Release Notes")
        alert.addButton(withTitle: "Close")
        switch alert.runModal() {
        case .alertFirstButtonReturn:
            NSPasteboard.general.clearContents()
            NSPasteboard.general.setString(command, forType: .string)
        case .alertSecondButtonReturn:
            if let url = URL(string: gitHubLatestReleaseURL) { NSWorkspace.shared.open(url) }
        default: break
        }
    }

    public class func showUpdateAlert(withUpdateURL url: URL?) {
        let command = "curl -s https://getappbox.com/install.sh | bash"
        let alert = NSAlert()
        alert.messageText = "New Version Available"
        alert.informativeText = "A newer version of \"AppBox\" is available.\n\nTo update, run this command in Terminal:"
        alert.alertStyle = .informational
        alert.accessoryView = commandTextField(command, width: 400)
        alert.addButton(withTitle: "Copy & Close")
        alert.addButton(withTitle: "Release Notes")
        alert.addButton(withTitle: "Close")
        switch alert.runModal() {
        case .alertFirstButtonReturn:
            NSPasteboard.general.clearContents()
            NSPasteboard.general.setString(command, forType: .string)
        case .alertSecondButtonReturn:
            if let releaseURL = URL(string: gitHubLatestReleaseURL) { NSWorkspace.shared.open(releaseURL) }
        default: break
        }
    }

    public class func showAlreadyUptoDateAlert() {
        let version = currentVersion ?? ""
        Common.showAlert(withTitle: "You\u{2019}re up-to-date!",
                         andMessage: "AppBox \(version) is currently the newest version available.")
    }

    /// Reports whether a newer version exists and, for direct (non-Homebrew) installs, the release URL.
    public class func isNewVersionAvailable(completion: @escaping (Bool, URL?) -> Void) {
        log.debug("Checking for new version...")
        let installedViaHomebrew = isInstalledViaHomebrew()
        let current = currentVersion
        Task {
            do {
                let latest = try await AppServices.serviceClient.fetchLatestVersion()
                if installedViaHomebrew {
                    let available = ABVersionCompare.isUpdateAvailable(latest: latest.homebrewVersion ?? latest.version,
                                                                       current: current)
                    await MainActor.run { completion(available, nil) }
                } else {
                    let available = ABVersionCompare.isUpdateAvailable(latest: latest.version, current: current)
                    await MainActor.run { completion(available, available ? latest.downloadURL : nil) }
                }
            } catch {
                log.info("Update check failed: \(error.localizedDescription, privacy: .public)")
                await MainActor.run { completion(false, nil) }
            }
        }
    }

    private class func commandTextField(_ command: String, width: CGFloat) -> NSView {
        let textField = NSTextField(frame: NSRect(x: 0, y: 0, width: width, height: 20))
        textField.stringValue = command
        textField.isEditable = false
        textField.isSelectable = true
        textField.isBordered = true
        textField.bezelStyle = .roundedBezel
        textField.font = NSFont.monospacedSystemFont(ofSize: 12.0, weight: .medium)
        textField.backgroundColor = .controlBackgroundColor
        textField.alignment = .center
        return textField
    }
}
